File: surf-server-selector-bukkit/surf-server-selector-bukkit-lobby/src/main/java/dev/slne/surf/surfserverselector/bukkit/lobby/listener/qole/DoubleJumpListener.java
package dev.slne.surf.surfserverselector.bukkit.lobby.listener.qole;

import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Particle;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerToggleFlightEvent;
import org.bukkit.util.Vector;

public final class DoubleJumpListener implements Listener {

  @EventHandler
  public void onPlayerMove(PlayerMoveEvent event) {
    final Player player = event.getPlayer();

    if (player.getGameMode() != GameMode.CREATIVE
        && player.getLocation().getBlock().getRelative(BlockFace.DOWN).getType() != Material.AIR
        && !player.isFlying()) {
      player.setAllowFlight(true);
    }
  }

  @EventHandler
  public void onPlayerToggleFlight(PlayerToggleFlightEvent event) {
    final Player player = event.getPlayer();

    if (player.getGameMode().equals(GameMode.CREATIVE)) {
      return;
    }

    final Vector direction = player.getEyeLocation().getDirection();
    final Location loc = player.getLocation();

    player.setVelocity(direction.multiply(2).setY(1));
    loc.getWorld().spawnParticle(Particle.EXPLOSION_NORMAL, loc, 10);

    player.setAllowFlight(false);
    player.setFlying(false);

    event.setCancelled(true);
  }
}

File: surf-lobby-paper/surf-lobby-paper-lobby/src/main/kotlin/dev/slne/surf/lobby/paper/lobby/features/DoubleJumpListener.kt
package dev.slne.surf.lobby.paper.lobby.features

import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.block.BlockFace
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.event.player.PlayerToggleFlightEvent
import org.springframework.stereotype.Component

@Component
class DoubleJumpListener : Listener {
    @EventHandler
    fun onPlayerMove(event: PlayerMoveEvent) {
        val player = event.getPlayer()

        if (player.gameMode != GameMode.CREATIVE && player.location.block.getRelative(BlockFace.DOWN).type != Material.AIR && !player.isFlying) {
            player.allowFlight = true
        }
    }

    @EventHandler
    fun onPlayerToggleFlight(event: PlayerToggleFlightEvent) {
        val player = event.getPlayer()

        if (player.gameMode == GameMode.CREATIVE) {
            return
        }

        val direction = player.eyeLocation.direction
        val loc = player.location

        player.velocity = direction.multiply(2).setY(1)
        loc.getWorld().spawnParticle(Particle.EXPLOSION, loc, 10)

        player.allowFlight = false
        player.isFlying = false

        event.isCancelled = true
    }
}